syntax = "proto3";

import "nanopb.proto";

package config;

message ConstantNode {
    repeated float values = 1 [(nanopb).max_count = 32];
}

message AdderNode {
    bool saturating = 1;
    uint32 inputs = 2;
}

message SubtractorNode {
    bool saturating = 1;
    uint32 inputs = 2;
}

message MultiplierNode {
    bool saturating = 1;
    uint32 inputs = 2;
}

message DividerNode {
    bool saturating = 1;
    uint32 inputs = 2;
}

message UniverseNode {
    message Light {
        string name = 1 [(nanopb).max_length = 64];
        repeated uint32 channels = 2 [(nanopb).max_count = 64];
    }

    repeated Light lights = 1 [(nanopb).max_count = 32];
    uint32 max_channel = 2;
}

message Node {
    string name = 1 [(nanopb).max_length = 64];

    oneof config {
        ConstantNode constant_node = 2;
        AdderNode adder_node = 3;
        SubtractorNode subtractor_node = 4;
        MultiplierNode multiplier_node = 5;
        DividerNode divider_node = 6;
        UniverseNode universe_node = 7;
    }
}

message Connection {
    uint32 from_node = 1;
    uint32 from_output = 2;
    uint32 to_node = 3;
    uint32 to_input = 4;
}

message Runner {
    repeated Node nodes = 1 [(nanopb).max_count = 128];
    repeated Connection connections = 2 [(nanopb).max_count = 128];
}
